stored_requests/ab_fetcher.go

```diff
@@ -0,0 +1,212 @@
+package stored_requests
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math/rand"
+
+	"github.com/buger/jsonparser"
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/prebid/prebid-server/metrics"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+// ABFetcher is an AllFetcher implementing the A/B experiment selection rules for stored requests only
+// The results from this fetcher MUST NOT be cached for the distribution rules to work.
+// The composite fetcher layout is ABFetcher -> FetcherWithCache -> Backend Fetcher (Files, HTTP, Db)
+// This fetcher SHOULD be in front of a caching fetcher to take advantage of caching on refetch.
+type ABFetcher struct {
+	fetcher       AllFetcher
+	metricsEngine metrics.MetricsEngine
+}
+
+// WithABFetcher returns an AllFetcher that may replace requested requests
+func WithABFetcher(fetcher AllFetcher, metricsEngine metrics.MetricsEngine) AllFetcher {
+	return &ABFetcher{
+		fetcher:       fetcher,
+		metricsEngine: metricsEngine,
+	}
+}
+
+// FetchRequests applies A/B experiment rules and refetches alternate stored requests if needed
+// triggered by ext.prebid.storedrequest.ab_config existence as a list of ABConfig objects
+func (f *ABFetcher) FetchRequests(ctx context.Context, requestIDs []string, impIDs []string) (requestData map[string]json.RawMessage, impData map[string]json.RawMessage, errs []error) {
+	if requestData, impData, errs = f.fetcher.FetchRequests(ctx, requestIDs, impIDs); len(errs) > 0 || len(requestIDs) != 1 || len(requestData) == 0 {
+		return
+	}
+	requestID := requestIDs[0]
+	storedRequest := requestData[requestID]
+
+	// Extract ab_config json if present
+	abConfigs, dataType, _, err := jsonparser.Get(storedRequest, "ext", openrtb_ext.PrebidExtKey, "storedrequest", "ab_config")
```

Comment:
Interesting approach. This wouldn't work for A/B Tests defined in the request sent to Prebid Server though, would it? Looks like it only works if the stored request itself defines the A/B Test. Would we want to allow publishers to define experiments in their Prebid.js config too?

Reply:
We limited the scope intentionally to just stored request A/B testing, to help with app config testing with its longer deploy timelines. Prebid.js can handle switching itself with better control that includes all bidders not just s2s.